Add mypy config and fix typing issues

michaelosthege opened the pull request:
This PR adds a `mypy` configuration that should help to diagnose and fix typing problems in the codebase.
(Related to #200, part of #800.)

With a [`mypy_groupby.py` script](https://gist.github.com/michaelosthege/24d0703e5f37850c9e5679f69598930a) the mypy results can be grouped by error code instead of location, making it easier to prioritize certain kinds of problems.

In addition to the configuration, I fixed a bunch of type hints, mostly those with `[has-type]`, `[valid-type]` and `[var-annotated]` error codes.

The majority of the fixes were only the annotations and got bundled in one commit.

For some fixes (like missing returns) I had to modify the code itself and because this could possibly effect behavior these went into their own commit.

--------------

Generally, we should use invariant hints like `Sequence[Variable]`/`Iterable[Variable]` instead of `List[Variable]` in function signatures.
That's because lists are invariant and such annotations will require callers to pass lists with the exact same type.
However, mypy doesn't warn about this per se.
After merging #799 it will be much easier to spot such issues from the PyMC/aeppl side.

michaelosthege commented on aesara/graph/basic.py:
```diff
@@ -662,7 +662,7 @@ def walk(
     return_children: bool = False,
     hash_fn: Callable[[T], Hashable] = id,
-) -> Generator[T, None, Dict[T, List[T]]]:
+) -> Generator[Union[T, Tuple[T, Optional[Sequence[T]]]], None, None]:
```
The order of items is `[YieldType, SendType, ReturnType]`

michaelosthege commented on aesara/graph/basic.py:
```diff
@@ -1679,8 +1679,9 @@ def expand(r):
-    results = ()
+    results: Tuple[Variable, ...] = ()
     for var in walk(graphs, expand, False):
+        assert isinstance(var, Variable)
```
The `walk` return type depends on the third argument, but `mypy` doesn't know this.
Either this `assert` or a `not isinstance(var, Variable): raise` tells `mypy` that henceforth `var: Variable` is correct.

ricardoV94 replied:
Can't we use cast?

https://docs.python.org/3/library/typing.html#typing.cast

michaelosthege replied:
Sound like a valid alternative: I don't have a preference.

michaelosthege commented on aesara/graph/fg.py:
```diff
@@ -227,12 +227,12 @@ def disown(self) -> None:
         self.variables = set()
-        self.inputs = None
-        self.outputs = None
+        self.inputs = []
+        self.outputs = []
```
They were initialized as `[]` in the `__init__` and this way we don't have to make them `Optional`.

michaelosthege commented on aesara/scan/op.py:
```diff
@@ -706,9 +706,7 @@ def tensorConstructor(shape, dtype):
-                typeConstructor(
-                    shape=(False,) + o.type.broadcastable, dtype=o.type.dtype
-                )
+                typeConstructor((False,) + o.type.broadcastable, o.type.dtype)
```
Because the signature hints the `TensorConstructorType` without specifying the arg names.

brandonwillard commented on aesara/tensor/__init__.py:
```diff
@@ -45,7 +45,7 @@ def as_tensor_variable(
 def _as_tensor_variable(
     x, name: Optional[str], ndim: Optional[int], **kwargs
-) -> NoReturn:
+) -> "TensorVariable":
```
What is the reason for this change?

michaelosthege replied:
The dispatch functions are confusing mypy. It looks like it's taking the type hint of this signature and expecting it for the dispatched functions that do return the value.
Kind of like with abstract functions I think.

brandonwillard commented on aesara/tensor/__init__.py:
```diff
@@ -80,7 +80,7 @@ def get_vector_length(v: Any):
 
 @singledispatch
-def _get_vector_length(op: Union[Op, Variable], var: Variable) -> NoReturn:
+def _get_vector_length(op: Union[Op, Variable], var: Variable):
```
Same here; why remove the type hint?

michaelosthege replied:
Here I couldn't figure out the correct type, but it's not `NoReturn` on the dispatched functions (similar to above).

The `NoReturn` was added in a PR that used monkeytype to infer types. Looks like it didn't do a great job for dispatch functions.

brandonwillard commented on aesara/tensor/subtensor.py:
```diff
@@ -135,8 +135,8 @@ def as_index_constant(
 def as_index_literal(
-    idx: Union[Variable, slice, type(np.newaxis)]
-) -> Union[int, slice, type(np.newaxis)]:
+    idx: Optional[Union[Variable, slice]]
+) -> Optional[Union[int, slice]]:
```
This doesn't seem necessary.

michaelosthege replied:
The `type(np.newaxis)` call is not allowed for a type annotation. It gives this error:
```
aesara\tensor\subtensor.py:138: error: Invalid type comment or annotation  [valid-type]
aesara\tensor\subtensor.py:138: note: Suggestion: use type[...] instead of type(...)
```
But with `type[np.newaxis]` there's another error:
```
aesara\tensor\subtensor.py:138: error: Variable "numpy.newaxis" is not valid as a type  [valid-type]
aesara\tensor\subtensor.py:138: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
```
That's probably because `np.newaxis is None`, so `Optional[Union[int, slice]]` remained.

brandonwillard replied:
What we actually need to do is evaluate `type(np.newaxis)` and use the result in the annotation, so `type[np.newaxis]` shouldn't make sense here.
 
It should be possible to evaluate that expression beforehand (e.g. in a `TYPE_CHECKING` block) and use the result in the annotation, but we can skip that for now and safely assume `np.newaxis` will be `None`.

michaelosthege commented on aesara/tensor/random/op.py:
```diff
@@ -27,9 +27,9 @@
     ndim_supp: int,
     dist_params: Sequence[Variable],
-    rep_param_idx: Optional[int] = 0,
+    rep_param_idx: int = 0,
```
`Optional` means that the argument can be `None`, which is not the case here.

michaelosthege commented on aesara/tensor/random/op.py:
```diff
@@ -27,9 +27,9 @@
+    rep_param_idx: int = 0,
     param_shapes: Optional[Sequence[Tuple[ScalarVariable]]] = None,
-) -> Tuple[int, ...]:
+) -> Union[TensorVariable, Tuple[ScalarVariable, ...]]:
```
The shape of a `Variable` is symbolic...

michaelosthege reviewed (commented):
@brandonwillard I rebased and fixed a conflict on a change where incorrect type hints were added on main (see my comments).
Can we get this PR merged?

brandonwillard commented on aesara/graph/op.py:
```diff
@@ -197,6 +197,10 @@ class Op(MetaObject):
 
     """
 
+    itypes = None
+    otypes = None
+    params_type: Optional[ParamsType] = None
+
```
This is not an exclusively type-information-based change; it could actually affect code that depends on `Op` (e.g. similar to how the `hasattr` checks below needed to be updated).

These kinds of changes need their own commits, if only to make the them stand out in some way.

michaelosthege replied:
I did that already, or do you want them to be split up even further?
![grafik](https://user-images.githubusercontent.com/5894642/154640705-e5c74eba-b3e1-4a11-9313-91938f8e9c13.png)

brandonwillard replied:
> I did that already, or do you want them to be split up even further?

Yes, each such independent change that affects a different class/area of code is eligible for its own commit, especially when each requires distinct considerations.

"Rectify typing with runtime code changes" is a catch-all that could encompass literally any and all substantial code changes, and such a themed commit requires reviewers to manually organize all the disparate changes before understanding what was really changed and (potentially) why. 

In contrast, other common catch-alls like "Refactor ..." are generally constrained to specific areas of the codebase and/or are assumed to contain no interface or functionality changes.  Those assumptions are much easier to check via a quick pass over the diffs.  

This commit doesn't have the above properties, which makes for a slower review cycle.

brandonwillard reviewed (changes requested):
These Mypy settings and type hint changes are not producing a passing `mypy` run locally.  More importantly, **we need a Mypy check in CI (e.g. via `pre-commit`) to validate/justify these changes**.

michaelosthege commented:
@brandonwillard I split the commit up as requested.

> These Mypy settings and type hint changes are not producing a passing mypy run locally. More importantly, we need a Mypy check in CI (e.g. via pre-commit) to validate/justify these changes.

Yes, and the Aesara project is at least months away from a passing mypy run unless you want to ignore/silence all these errors.
I personally have **neither the motivation nor the time** to do all this work while rebasing the branch over and over again.

Since pre-commit doesn't print the output of passing hooks, it doesn't help to have a mypy step with an always-0 exit code.
However, I could add a mypy job to the CI that prints everything but doesn't fail the CI on errors.
Would that be a satisfactory strategy?

michaelosthege commented:
@brandonwillard I added a script to run mypy and compare the results against a list of modules known to pass.
It raises errors on regressions, but also on unexpected successes.
I added it to the pre-commit too.

Note that fixing type error sometimes leads to _new_ errors and there's no guarantee this won't happen to files already on the list.

brandonwillard reviewed (approved):
I've updated the Mypy config and pre-commit setup to one that simply enumerates the modules that need to be fixed and ignores their errors.  We can start removing those config entries as we fix and add type hints throughout the library.

As long as this passes, we can merge it.
